collect in vector runtime causes "panic: collection: partial not an array type"

## tl;dr

With input `data.sup`:

```
{a:1,b:"one"}
{a:2,b:"two"}
```

This query causes a panic:

```
$ SUPER_VAM=1 super -c "from data.sup | foo := collect(a) by b"
panic: collection: partial not an array type
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:26 +0x5e
github.com/brimdata/super/runtime/sam/op.(*Catcher).Pull.func1()
	/Users/phil/work/super/runtime/sam/op/catcher.go:25 +0x3d
panic({0xb153a60?, 0xb499dd0?})
	/usr/local/opt/go/libexec/src/runtime/panic.go:792 +0x132
github.com/brimdata/super/runtime/vam/expr/agg.(*collect).ConsumeAsPartial(0xb4baa68?, {0xb4bb068?, 0xc0001840f0?})
	/Users/phil/work/super/runtime/vam/expr/agg/collect.go:42 +0x205
github.com/brimdata/super/runtime/vam/op/aggregate.(*superTable).update(0xc00053a000, {0xc0001aa000, 0x1, 0x2}, {0xc0001aa010, 0x1, 0x9520f85?})
	/Users/phil/work/super/runtime/vam/op/aggregate/aggtable.go:68 +0x506
github.com/brimdata/super/runtime/vam/op/aggregate.(*Aggregate).consume(0xc00046bad0, {0xc0001aa000, 0x1, 0x2}, {0xc0001aa010, 0x1, 0x1})
	/Users/phil/work/super/runtime/vam/op/aggregate/aggregate.go:103 +0x27e
github.com/brimdata/super/runtime/vam/op/aggregate.(*Aggregate).Pull.func1({0xc0001aa000, 0x2, 0xc0000316b0?})
	/Users/phil/work/super/runtime/vam/op/aggregate/aggregate.go:84 +0x7b
github.com/brimdata/super/vector.Apply(0x10?, 0xb4baa08?, {0xc0001aa000?, 0xc000031738?, 0xb279080?})
	/Users/phil/work/super/vector/apply.go:17 +0xa3
github.com/brimdata/super/runtime/vam/op/aggregate.(*Aggregate).Pull(0xc00046bad0, 0x0?)
	/Users/phil/work/super/runtime/vam/op/aggregate/aggregate.go:83 +0x13b
github.com/brimdata/super/runtime/vam.(*Materializer).Pull(0xc0004a9ba0?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:26 +0x27
github.com/brimdata/super/runtime/sam/op.(*Single).Pull(0xc0004cba40, 0x5?)
	/Users/phil/work/super/runtime/sam/op/mux.go:120 +0x33
github.com/brimdata/super/runtime/sam/op.(*Catcher).Pull(0x83332e01c682a8af?, 0x79?)
	/Users/phil/work/super/runtime/sam/op/catcher.go:28 +0x5c
github.com/brimdata/super/runtime/exec.(*Query).Pull(0x950e0df?, 0x40?)
	/Users/phil/work/super/runtime/exec/query.go:49 +0x3c
github.com/brimdata/super/zbuf.CopyMux(0xc000031ca8, {0xb4a6c80, 0xc0004cba70})
	/Users/phil/work/super/zbuf/mux.go:40 +0x38
github.com/brimdata/super/cmd/super/root.(*Command).Run(0xc00042e600, {0xc00003c710, 0x0, 0x0})
	/Users/phil/work/super/cmd/super/root/command.go:167 +0x925
github.com/brimdata/super/pkg/charm.path.run({0xc000080c00, 0x1, 0x1}, {0xc00003c710, 0x0, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7b
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0xc5f3380, {0xc00003c700, 0x2, 0x2})
	/Users/phil/work/super/pkg/charm/charm.go:71 +0x105
main.main()
	/Users/phil/work/super/cmd/super/main.go:40 +0x5b
```

## Details

Repro is with super commit 7473504. This is a simplification of an issue originally found in a "crunchy query" provided by a community zync user.

The panic does not occur in sequential runtime.

```
$ super -version
Version: v1.18.0-396-g747350475

$ super -c "from data.sup | foo := collect(a) by b"
{b:"one",foo:[1]}
{b:"two",foo:[2]}
```

--- a/runtime/vam/expr/agg/collect.go
+++ b/runtime/vam/expr/agg/collect.go
@@ -37,14 +37,23 @@ func (c *collect) ConsumeAsPartial(partial vector.Any) {
 	if c, ok := partial.(*vector.Const); ok && c.Value().IsNull() {
 		return
 	}
+	n := partial.Len()
+	var index []uint32
+	if view, ok := partial.(*vector.View); ok {
+		partial, index = view.Any, view.Index
+	}
 	array, ok := partial.(*vector.Array)
 	if !ok {
 		panic("collection: partial not an array type")
 	}
 	var b zcode.Builder
 	typ := array.Values.Type()
-	for i := range array.Len() {
-		for k := array.Offsets[i]; k < array.Offsets[i+1]; k++ {
+	for i := range n {
+		idx := i
+		if index != nil {
+			idx = index[i]
+		}
+		for k := array.Offsets[idx]; k < array.Offsets[idx+1]; k++ {
 			b.Truncate()
 			array.Values.Serialize(&b, k)
 			c.samcollect.Consume(super.NewValue(typ, b.Bytes().Body()))

--- a/runtime/ztests/op/aggregate/collect-partial-view.yaml
+++ b/runtime/ztests/op/aggregate/collect-partial-view.yaml
@@ -1,0 +1,18 @@
+# Tests that collect correctly handles partial array values that are wrapped
+# in a view.
+script: |
+  super -z -c 'from "test.sup" | collect(a) by b | sort b'
+
+vector: true
+
+inputs:
+  - name: test.sup
+    data: |
+      {a:1,b:"one"}
+      {a:2,b:"two"}
+
+outputs:
+  - name: stdout
+    data: |
+      {b:"one",collect:[1]}
+      {b:"two",collect:[2]}

